crates/turbo-tasks-memory/src/concurrent_priority_queue.rs

```diff
@@ -0,0 +1,171 @@
+use std::{
+    cmp::min,
+    collections::{hash_map::RandomState, BinaryHeap},
+    fmt::Debug,
+    hash::{BuildHasher, Hash, Hasher},
+};
+
+use once_cell::sync::OnceCell;
+use parking_lot::{Mutex, MutexGuard};
+use priority_queue::PriorityQueue;
+
+fn shard_amount() -> usize {
+    static SHARD_AMOUNT: OnceCell<usize> = OnceCell::new();
+    *SHARD_AMOUNT.get_or_init(|| {
+        (std::thread::available_parallelism().map_or(1, usize::from) * 16).next_power_of_two()
+    })
+}
+
+fn shift() -> usize {
+    static SHIFT: OnceCell<usize> = OnceCell::new();
+    *SHIFT
+        .get_or_init(|| std::mem::size_of::<usize>() * 8 - shard_amount().trailing_zeros() as usize)
+}
+
+pub struct ConcurrentPriorityQueue<K: Hash + Eq, V: Ord, H = RandomState> {
+    shards: Box<[Mutex<PriorityQueue<K, V, H>>]>,
+    hasher: H,
+}
+
+impl<K: Hash + Eq, V: Ord + Clone + Debug, H: BuildHasher + Default + Clone>
+    ConcurrentPriorityQueue<K, V, H>
+{
+    pub fn new() -> Self {
+        Self::with_hasher(Default::default())
+    }
+}
+
+impl<K: Hash + Eq, V: Ord + Clone + Debug, H: BuildHasher + Clone>
+    ConcurrentPriorityQueue<K, V, H>
+{
+    pub fn with_hasher(hasher: H) -> Self {
+        let shards = (0..shard_amount())
+            .map(|_| Mutex::new(PriorityQueue::with_hasher(hasher.clone())))
+            .collect::<Vec<_>>();
+        Self {
+            shards: shards.into_boxed_slice(),
+            hasher,
+        }
+    }
+
+    fn shard(&self, key: &K) -> MutexGuard<PriorityQueue<K, V, H>> {
+        let mut hash = self.hasher.build_hasher();
+        key.hash(&mut hash);
+        let index = ((hash.finish() as usize) << 7) >> shift();
+        unsafe { self.shards.get_unchecked(index) }.lock()
+    }
+
+    #[allow(dead_code)]
+    pub fn len(&self) -> usize {
+        let mut len = 0;
+        for shard in self.shards.iter() {
+            len += shard.lock().len()
+        }
+        len
+    }
+
+    pub fn insert(&self, key: K, value: V) {
+        let mut inner = self.shard(&key);
+        inner.push(key, value);
+    }
+
+    pub fn upsert(
```

Comment:
Feels like this would be closer to the std APIs:
```suggestion
    pub fn upsert_with(
```